File: packages/go/whatsapp-go/metering_test.go
package whatsapp

import (
	"testing"
	"time"
)

func TestFranquiaService(t *testing.T) {
	if BillableUnits(0, 1000) != 0 || BillableUnits(1000, 1000) != 0 {
		t.Fatal("franquia deveria zerar")
	}
	if BillableUnits(1001, 1000) != 1 || BillableUnits(5000, 1000) != 4000 {
		t.Fatal("billable inesperado")
	}
	b, f := BillableDeltaForIncrement(950, 100, 1000)
	if b != 50 || f != 50 {
		t.Fatalf("delta atraves da franquia: %d %d", b, f)
	}
}

func TestMeteringAcumula(t *testing.T) {
	clock := func() time.Time { return time.Date(2026, 9, 16, 12, 0, 0, 0, time.UTC) }
	m := NewMeteringService(CenarioBasePlan, NewInMemoryUsageStore(), clock, "America/Sao_Paulo")
	d1, err := m.Consume("tenant-a", CategoryAuth, 3)
	if err != nil {
		t.Fatal(err)
	}
	if d1.Period != "2026-09" || d1.Used != 3 || d1.BillableDelta != 3 {
		t.Fatalf("%+v", d1)
	}
	if _, err := m.Consume("tenant-a", CategoryUtility, 1); err != nil {
		t.Fatal(err)
	}
	snap, _ := m.Snapshot("tenant-a", "")
	if snap.Counts.Auth != 3 || snap.Counts.Utility != 1 {
		t.Fatalf("%+v", snap.Counts)
	}
}

func TestMeteringHardNaoPersiste(t *testing.T) {
	quota := int64(2)
	plan := MergePlan(&PlanOverride{
		QuotaMode: QuotaHard,
		Categories: map[MessageCategory]CategoryQuotaPatch{
			CategoryAuth: {MonthlyQuota: &quota},
		},
	}, QuotaHard)
	clock := func() time.Time { return time.Date(2026, 9, 16, 12, 0, 0, 0, time.UTC) }
	m := NewMeteringService(plan, NewInMemoryUsageStore(), clock, "")
	if _, err := m.Consume("t", CategoryAuth, 2); err != nil {
		t.Fatal(err)
	}
	_, err := m.Consume("t", CategoryAuth, 1)
	if _, ok := err.(*QuotaExceededError); !ok {
		t.Fatalf("esperado QuotaExceededError, got %v", err)
	}
	snap, _ := m.Snapshot("t", "")
	if snap.Counts.Auth != 2 {
		t.Fatalf("persistiu excedente: %d", snap.Counts.Auth)
	}
}

func TestMeteringSoftOverage(t *testing.T) {
	quota := int64(2)
	plan := MergePlan(&PlanOverride{
		QuotaMode: QuotaSoft,
		Categories: map[MessageCategory]CategoryQuotaPatch{
			CategoryMarketing: {MonthlyQuota: &quota},
		},
	}, QuotaSoft)
	clock := func() time.Time { return time.Date(2026, 9, 16, 12, 0, 0, 0, time.UTC) }
	m := NewMeteringService(plan, NewInMemoryUsageStore(), clock, "")
	if _, err := m.Consume("t", CategoryMarketing, 2); err != nil {
		t.Fatal(err)
	}
	over, err := m.Consume("t", CategoryMarketing, 1)
	if err != nil {
		t.Fatal(err)
	}
	if !over.Allowed || !over.Overage || over.OverageDelta != 1 || over.Used != 3 {
		t.Fatalf("%+v", over)
	}
}

func TestServiceFaturaApos1000(t *testing.T) {
	clock := func() time.Time { return time.Date(2026, 9, 16, 12, 0, 0, 0, time.UTC) }
	m := NewMeteringService(CenarioBasePlan, NewInMemoryUsageStore(), clock, "")
	first, err := m.Consume("t", CategoryService, 1000)
	if err != nil {
		t.Fatal(err)
	}
	if first.BillableDelta != 0 || first.FreeDelta != 1000 {
		t.Fatalf("%+v", first)
	}
	next, err := m.Consume("t", CategoryService, 5)
	if err != nil {
		t.Fatal(err)
	}
	if next.BillableDelta != 5 || next.Used != 1005 {
		t.Fatalf("%+v", next)
	}
}

File: packages/nodejs/whatsapp-sdk/tests/metering.test.ts
import { describe, expect, it } from "vitest";
import { QuotaExceededError } from "../src/errors.js";
import {
  billableDeltaForIncrement,
  billableUnits,
  InMemoryUsageStore,
  MeteringService,
} from "../src/metering.js";
import { CENARIO_BASE_PLAN, mergePlan } from "../src/plan.js";

describe("franquia Service (1.000 sessões/mês)", () => {
  it("nao fatura os primeiros 1000", () => {
    expect(billableUnits(0, 1_000)).toBe(0);
    expect(billableUnits(1_000, 1_000)).toBe(0);
    expect(billableUnits(1_001, 1_000)).toBe(1);
    expect(billableUnits(5_000, 1_000)).toBe(4_000);
  });

  it("reparte delta que atravessa a franquia", () => {
    expect(billableDeltaForIncrement(950, 100, 1_000)).toEqual({
      billableDelta: 50,
      freeDelta: 50,
    });
    expect(billableDeltaForIncrement(0, 500, 1_000)).toEqual({
      billableDelta: 0,
      freeDelta: 500,
    });
    expect(billableDeltaForIncrement(1_000, 10, 1_000)).toEqual({
      billableDelta: 10,
      freeDelta: 0,
    });
  });
});

describe("MeteringService", () => {
  const clock = () => new Date("2026-09-16T12:00:00Z");

  it("acumula por categoria e periodo YYYY-MM", async () => {
    const metering = new MeteringService(CENARIO_BASE_PLAN, new InMemoryUsageStore(), clock);
    const d1 = await metering.consume("tenant-a", "auth", 3);
    const d2 = await metering.consume("tenant-a", "utility", 1);
    expect(d1.period).toBe("2026-09");
    expect(d1.used).toBe(3);
    expect(d1.billableDelta).toBe(3);
    expect(d2.category).toBe("utility");
    const snap = await metering.snapshot("tenant-a");
    expect(snap.counts).toEqual({ auth: 3, utility: 1, service: 0, marketing: 0 });
  });

  it("modo hard bloqueia e nao persiste o consumo excedente", async () => {
    const plan = mergePlan({
      quotaMode: "hard",
      categories: { auth: { monthlyQuota: 2, unitCostUsd: 0.0315, freeAllowance: 0 } },
    });
    const metering = new MeteringService(plan, new InMemoryUsageStore(), clock);
    await metering.consume("t", "auth", 2);
    await expect(metering.consume("t", "auth", 1)).rejects.toBeInstanceOf(QuotaExceededError);
    const snap = await metering.snapshot("t");
    expect(snap.counts.auth).toBe(2);
  });

  it("modo soft permite envio e marca overage", async () => {
    const plan = mergePlan({
      quotaMode: "soft",
      categories: { marketing: { monthlyQuota: 2, unitCostUsd: 0.0625, freeAllowance: 0 } },
    });
    const metering = new MeteringService(plan, new InMemoryUsageStore(), clock);
    await metering.consume("t", "marketing", 2);
    const over = await metering.consume("t", "marketing", 1);
    expect(over.allowed).toBe(true);
    expect(over.overage).toBe(true);
    expect(over.overageDelta).toBe(1);
    expect(over.used).toBe(3);
    expect(over.remaining).toBe(0);
  });

  it("service so comeca a faturar apos 1000", async () => {
    const metering = new MeteringService(CENARIO_BASE_PLAN, new InMemoryUsageStore(), clock);
    const first = await metering.consume("t", "service", 1_000);
    expect(first.billableDelta).toBe(0);
    expect(first.freeDelta).toBe(1_000);
    expect(first.overage).toBe(false);
    const next = await metering.consume("t", "service", 5);
    expect(next.billableDelta).toBe(5);
    expect(next.freeDelta).toBe(0);
    expect(next.used).toBe(1_005);
  });
});
